lighthouse-core/gather/gatherers/form-elements.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+
+/**
+ *  @param {HTMLFormElement} formElement
+ *  @return {Array<HTMLElement>}
+ */
+/* istanbul ignore next */
+function getChildrenInputs(formElement) {
+  if (formElement.formless == true){
+    return formElement.inputs
+  }
+
+  const inputsArray = [];
+  const childrenArray = Array.prototype.slice.call(formElement.childNodes);
+  
+  childrenArray.map(  /** @param {HTMLElement} node */ (element) => {
+    if (element.nodeName == 'INPUT' || element.nodeName == 'SELECT' || element.nodeName == 'TEXTAREA'){
+      const inputAttributes = {
+        id: element.id,
+        nodeName: element.nodeName,
+        name: element.name,
+        placeholder: element.placeholder,
+        autocomplete: element.autocomplete,
+      }
+      inputsArray.push(inputAttributes);
+    }
+  });
+
+  return inputsArray;
+}
+
+/**
+ *  @param {HTMLFormElement} formElement
+ *  @return {Array<HTMLElement>}
+ */
+/* istanbul ignore next */
+function getChildrenLabels(formElement) {
+  if (formElement.formless == true){
+    return formElement.labels
+  }
+  const childrenArray = Array.prototype.slice.call(formElement.childNodes);
+  const labelsArray = [];
+  childrenArray.map(/** @param {HTMLElement} node */ (element) => {
+    if (element.nodeName == 'LABEL'){
+      const labelAttributes = {
+        id: element.id,
+        nodeName: element.nodeName,
+        name: element.name,
+        for: element.for,
+      }
+      labelsArray.push(labelAttributes);
+    }
+  });
+
+  return labelsArray;
+}
+
+
+/**
+ * @return {LH.Artifacts['FormElements']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+
+  const formElements = getElementsInDocument('form'); // eslint-disable-line no-undef
+  const formChildren = getElementsInDocument('textarea', 'input', 'label', 'select'); // eslint-disable-line no-undef
+
+  const formless = {
+    formless: true,
+    inputs: [],
+    labels: [],
+  }
+  formChildren.map(/** @param {HTMLElement} node */ (childElement) => {
+    if (childElement.form == "" || !childElement.form){
```

Comment:
some styling nits for readability:

1)

I think it can just be 
```
if (!childElement.form) { ...
```

2) for if statements that span an entire block, we like to use early exits:

```
if (childElement.form) return;
```

3) since you aren't using the _result_ of this `.map`, prefer iteration using `for (const childElement of formChildren) { ...` (and then use `continue` instead of `return` from point 2)

Reply:
ok cool, let me go and change that.
Also what about the forEach function? would that be suitbale too?

Reply:
`for ... of` is preferred: first-class language constructs are better than using a function to do the same thing.

Reply:
makes sense, got it!